packages/flutter/lib/src/rendering/custom_layout.dart

```diff
@@ -20,12 +20,60 @@ class MultiChildLayoutParentData extends ContainerBoxParentDataMixin<RenderBox>
 
 /// A delegate that controls the layout of multiple children.
 ///
+/// Delegates must be immutable. Specifically, if two delegates are equal, then
+/// then must produce the same layout. To change the layout, replace the
+/// delegate with a different instance.
+///
+/// Override [getSize] to control the overall size of the layout. The size of
+/// the layout cannot depend on layout properties of the children.
+///
+/// Override [performLayout] to size and position the children. An
+/// implementation of [performLayout] must call [layoutChild] exactly once for
+/// each child, but it may call [layoutChild] on children in an arbitrary order.
+/// Typically a delegate will use the size returned from [layoutChild] on one
+/// child to determine the constraints for [performLayout] on another child or
+/// to determine the offset for that child or another child.
+///
+/// Override [shouldRelayout] to determine when the layout of the children needs
+/// to be recomputed when the delegate changes.
```

Comment:
maybe we should document and even assert that x.shouldRelayout(x) returns false?